### DISCUSSION TITLE
Lora SX1280 with FHSS

### DISCUSSION THREAD

**Discussion by HamzaYslmn** [jgromes/RadioLib#1348] [Q&A]
I want to use FHSS using sx1280 and make a two-way custom RC application. But I couldn't find a library or example for this. I need help.

**Comment by HeadBoffin** [comment]
Searching the discussions will explain why there are no examples and why there won't be.

**Reply by HamzaYslmn** [reply]
> Searching the discussions will explain why there are no examples and why there won't be.

![image](https://github.com/user-attachments/assets/49875404-5b14-4fc3-9e89-c80aaebaff81)


I am the only result with these keywords.

If I use ELRS Airport, it has FHSS support. But using ELRS will be an extra device cost for me.

**Comment by jgromes** [comment]
@HamzaYslmn What leads you to believe SX1280 is capable of transmitting FHSS? There is no mention of this modulation in the device datasheet.

**Reply by HamzaYslmn** [reply]
> @HamzaYslmn What leads you to believe SX1280 is capable of transmitting FHSS? There is no mention of this modulation in the device datasheet.

I thought FHSS support could be provided in software, as in the ELRS example.

nrf24L01 does not have FHSS support in its datasheet, but I provided this support using RF24 library.
https://github.com/HamzaYslmn/NRFFHSS-ESP32

**Reply by jgromes** [reply]
Depends on what you mean by FHSS. The example you have linked seems to change the frequency between frames, not within a frame. All references to FHSS in this library (FHSS on SX127x or LR-FHSS on SX126x) refer to hopping frequency while transitting a single frame. I don't think that's something the SX128x can do.

You are free to change the frequency between subsequent transmissions (within legal limits of course) usign the `setFrequency` method, but that's completely within the scope of the user code.

**Reply by HamzaYslmn** [reply]
> Depends on what you mean by FHSS. The example you have linked seems to change the frequency between frames, not within a frame. All references to FHSS in this library (FHSS on SX127x or LR-FHSS on SX126x) refer to hopping frequency while transitting a single frame. I don't think that's something the SX128x can do.
> 
> You are free to change the frequency between subsequent transmissions (within legal limits of course) usign the `setFrequency` method, but that's completely within the scope of the user code.



I was wondering if you provide software-based support for FHSS, or if there are any libraries that offer such support.
https://github.com/ExpressLRS/ExpressLRS because ELRS does.

But when I use ELRS, one of my devices is completely dedicated to ELRS, I cannot add anything extra.
So I am looking for a solution.

FHSS support is not mandatory, but there may be up to 20 devices operating side by side on the same channel. Additionally, there will be Wi-Fi and Bluetooth signals in the environment. My goal is to minimize interference and maximum signal penetration as much as possible.

**Reply by jgromes** [reply]
Please read my previous post again - what do you mean by FHSS? Hopping while transmitting a frame, or hopping between frames?

The first one requires support from the radio hardware, which I don't know if SX128x is capable of. The second one is up to the user.

**Reply by HamzaYslmn** [reply]
> Please read my previous post again - what do you mean by FHSS? Hopping while transmitting a frame, or hopping between frames?
> 
> The first one requires support from the radio hardware, which I don't know if SX128x is capable of. The second one is up to the user.

I am looking for a solution that supports 2nd. 
Feature 1 should be supported on the hardware side as you said.

**Reply by jgromes** [reply]
In that case it's like I wrote - that's up to you. This library is primarly a driver, it provides interface to the features of the radio hardware (and abstracts away which exact radio hardware you are using). You can call `setFrequency` between your frames as required by any kind of hopping algorithm you define. Not sure what additional support you need from the library to implement your frequency hopping.

**Reply by HamzaYslmn** [reply]
> In that case it's like I wrote - that's up to you. This library is primarly a driver, it provides interface to the features of the radio hardware (and abstracts away which exact radio hardware you are using). You can call `setFrequency` between your frames as required by any kind of hopping algorithm you define. Not sure what additional support you need from the library to implement your frequency hopping.

I am currently looking for a pre-written module or library that can handle this. Managing time synchronization is exhausting for me, and I find it quite difficult to grasp.

**Reply by jgromes** [reply]
I'm fairly certain you won't find it, but it's your life - do what you like with it ;)

**Reply by HamzaYslmn** [reply]
If I can't find it, I'll use ELRS, it supports it anyway :)
